R/mwrite.R

```diff
@@ -0,0 +1,415 @@
+.quotes <- function(x) paste0("\"", x, "\"")
+
+tocode <- function(l) {
+  if(!length(l)) return(character(0))
+  sc <- sapply(l, function(x) is.character(x) && length(x)==1)
+  l[sc] <- sapply(l[sc], .quotes)
+  paste0(names(l), " = ", as.character(l))
+}
+
+# block name is $OMEGA or $SIGMA
+mwrite_matrix <- function(x, block_name) {
+  code <- character(0)
+  for(i in seq_along(x$data)) {
+    datai <-  x$data[[i]]
+    labelsi <- x$labels[[i]]
+    namei <- x$names[[i]]
+    code <- c(code, block_name)
+    header <- "@block"
+    if(namei != "...") {
+      header <- c(header, paste0("@name ", namei))  
+    }
+    if(any(labelsi != ".")) {
+      labels_ <- paste0(labelsi, collapse = " ")
+      header <- c(header, paste0("@labels ", labels_))
+    } 
+    code <- c(code, header)
+    for(i in seq_along(datai)) {
+      tag <- paste0("// row ", i)
+      code <- c(code,  tag, datai[[i]])  
+    }
+    code <- c(code, "")
+  }
+  code
+}
+
+get_upper_tri <- function(x) {
+  x <- as.matrix(x)
+  n <- nrow(x)
+  x <- x[upper.tri(x, diag = TRUE)]
+  l <- vector(mode = "list", length = n)
+  w <- 1
+  for(i in seq(n)) {
+    s <- seq(w, w+i-1)
+    l[[i]] <- x[s]
+    w <- w + i
+  }
+  names(l) <- paste0("row", seq(n))
+  l
+}
+
+require_yaml <- function() {
+  if(!requireNamespace("yaml", quietly = TRUE)) {
+    abort("The package \"yaml\" is required.") # nocov
+  } 
+}
+
+mwrite_model_to_list <- function(x) {
+  l <- list()
+  # Header
+  l$source <- "mrgsolve::mwrite"
+  l$mrgsolve <- as.character(packageVersion("mrgsolve"))
+  l$format <- "list"
+  l$version <- 1
+  l$model <- x@model
+  # Problem
+  l$prob <- ""
+  # Critical items
+  l$param <- as.list(param(x))
+  l$init <- as.list(init(x))
+  l$capture <- character(0)
+  if(length(x@capture)) {
+    capture <- names(x@capture)
+    new_names <- unname(x@capture)
+    same_name <- capture == new_names
+    l$capture <- paste0(new_names, " = ", capture)
+    l$capture[same_name] <- capture[same_name]
+  }
+  # Omega and Sigma
+  l$omega <- list()
+  l$omega$data <- lapply(as.list(omat(x)), get_upper_tri)
+  l$omega$labels <- labels(omat(x))
+  l$omega$names <- names(omat(x))
+  if(length(l$omega$data)) {
+    names(l$omega$data) <- paste0("matrix", seq_along(l$omega$data))
+    names(l$omega$labels) <- paste0("matrix",seq_along(l$omega$labels))
+  }
+  l$sigma <- list()
+  l$sigma$data <- lapply(as.list(smat(x)), get_upper_tri)
+  l$sigma$labels <- labels(smat(x))
+  l$sigma$names <- names(smat(x))
+  if(length(l$sigma$data)) {
+    names(l$sigma$data) <- paste0("matrix", seq_along(l$sigma$data))
+    names(l$sigma$labels) <- paste0("matrix",seq_along(l$sigma$labels))
+  }
+  # Other
+  l$env <- as.list(x@envir)
```

Comment:
Would `envir` be better for consistency?

Reply:
Good call.